### PR TITLE
[PAC] Implement authentication for C++ member function pointers

### PR DESCRIPTION
**Pull request by ojhunt**
Introduces type based signing of member function pointers. To support this discrimination schema we no longer emit member function pointer to virtual methods and indices into a vtable but migrate to using thunks. This does mean member function pointers are no longer necessarily directly comparable, however as such comparisons are UB this is acceptable.

We derive the discriminator from the C++ mangling of the type of the pointer being authenticated.

Co-Authored-By: Akira Hatanaka ahatanaka@apple.com
Co-Authored-By: John McCall rjmccall@apple.com

### CODE REVIEW

**Comment by ojhunt** [comment]
@asl could you check this as well? (I apparently can't add/request reviewers?)

File: clang/test/CodeGenCXX/ptrauth-member-function-pointer.cpp
Comment (kovdan01) [inline clang/test/CodeGenCXX/ptrauth-member-function-pointer.cpp:8]: Nit: it's probably worth to re-order C++ sources and/or CHECK lines in a way that CHECK lines are close to corresponding source code lines. For example, `gmethod*` are defined in the next "code section" (while it's possible to define them here), but tested here. It looks like that we can't achive such "test locality" for all the definitions, but some enhancements seem possible.

However, such change might not be very trivial, and I'm OK with postponing it to the next tests enhancements patch - I'll submit a one adding checks against ELF triples anyway.

File: clang/lib/CodeGen/CGPointerAuth.cpp
Comment (kovdan01) [inline clang/lib/CodeGen/CGPointerAuth.cpp:370]: Nit

```suggestion
  const auto &Schema = getCodeGenOpts().PointerAuth.CXXMemberFunctionPointers;
```

File: clang/lib/CodeGen/ItaniumCXXABI.cpp
Comment (kovdan01) [inline clang/lib/CodeGen/ItaniumCXXABI.cpp:884]: Nit

```suggestion
  const auto *CPA = dyn_cast<llvm::ConstantPtrAuth>(Ptr);
```

File: clang/lib/CodeGen/ItaniumCXXABI.cpp
Comment (kovdan01) [inline clang/lib/CodeGen/ItaniumCXXABI.cpp:1180]: Could you please explain the rationale for such a solution? It's probably worth adding more detailed comments here and in other places where the LSB of adj value is significant.

**Review by kovdan01 — COMMENTED** [review]
LGTM with minor non-blocking nits. However, I'd prefer to wait for @asl 's review before merging the PR. I'm OK with merging whenever @asl is.

File: clang/test/Preprocessor/ptrauth_feature.c
Comment (ojhunt) [reply clang/test/Preprocessor/ptrauth_feature.c:74]: Weird mismerge I guess?

- [ ] fix the merge

File: clang/lib/CodeGen/ItaniumCXXABI.cpp
Comment (ojhunt) [reply clang/lib/CodeGen/ItaniumCXXABI.cpp:1180]: @kovdan01 what do you mean here? the use of a thunk in general?

File: clang/test/Preprocessor/ptrauth_feature.c
Comment (kovdan01) [reply clang/test/Preprocessor/ptrauth_feature.c:74]: > Weird mismerge I guess?

Yep, I guess so

File: clang/lib/CodeGen/ItaniumCXXABI.cpp
Comment (kovdan01) [reply clang/lib/CodeGen/ItaniumCXXABI.cpp:1180]: > @kovdan01 what do you mean here? the use of a thunk in general?

I'm not sure that I correctly understand the logic behind not setting the LSB of adj to 1 with pauth for member function pointers enabled - see the comment in code. What's the purpose of this least significant bit value and how and where it's interpreted after being set here?

Comment (rjmccall) [reply clang/lib/CodeGen/ItaniumCXXABI.cpp:1180]: The LSB of the offset indicates whether the stored function value is an offset within the v-table (normally used for virtual member functions) or a function pointer (normally used for non-virtual member functions).  Because we are emitting the member function pointer with a function pointer to a thunk, we must clear the bit, or else calls will try to interpret the function pointer as an offset within the v-table.

Comment (kovdan01) [reply clang/lib/CodeGen/ItaniumCXXABI.cpp:1180]: Thanks @rjmccall for explanation!

**Review by asl — APPROVED** [review]
Thanks!

**Comment by asl** [comment]
Linux BuildKite failure seems to be completely unrelated (likely due to #99930)

**Pull request merged**
